### ISSUE TITLE
feature suggestion | Synchronous/Asynchronous Helpers

### ISSUE THREAD

**Issue by robincsamuel** [wycats/handlebars.js#717]
Registering a helper as Sync or Async, It helps to listen for callbacks, and get the data from callback.

**Comment by nknapp** [comment]
I have found [the handlebars-async](https://www.npmjs.com/package/handlebars-async) package on npm. But it's a little older and I don't know if it works with the current Handlebars-version.

I have also just written something similar for promises. The package [promised-handlebars](https://www.npmjs.com/package/promised-handlebars) allows you to return promises from within helpers. I plan to use it in one of my projects, but so far it hasn't been used in a production environment. But there  are unit tests for several few edge-cases (such as calling async-helpers from within async block-helpers) and they are all green...

**Comment by ErisDS** [comment]
@nknapp that sounds amazing! [express-hbs](https://github.com/barc/express-hbs) has async support, and the async works for block helpers, but nesting async helpers does not work - so I'm really interested to see this working - means there's hope yet for express-hbs :+1:

**Comment by nknapp** [comment]
@ErisDS, do you think I should post it there. I didn't now that `express-hbs` couldn't  nest async helper. My primary focus is not `express`, but a README-generator I'm currently working on. I would really appreciate other people to try it ([promised-handlebars](https://www.npmjs.com/package/promised-handlebars)) give feedback.

**Comment by ElChupacabra26** [comment]
To add to the valid use use-cases, what if you need to fetch values from a translation DB based on current locale?

```handlebars
<div class="howItWorks">
    {{{i18nFetch id=how-it-works locale=locale}}}
</div>
```

Moreover, what about adding a CMS block from a DB entry using a dynamic id like:

```handlebars
<div class="searchCms">
    {{{cmsLoader 'search-{$term}' term=params.input defaultId='search-default'}}}
</div>
```

This is especially useful for server-side rendering (i.e. using [express-handlebars](https://github.com/ericf/express-handlebars)).

**Comment by dwhieb** [comment]
Here's another use case: I'm writing a documentation-generator for Swagger ([simple-swagger](https://github.com/dwhieb/simple-swagger)), which allows for external schema definitions. I'd like to write a Handlebars helper that recognizes when a schema is defined externally, goes to the provided URL where that schema lives, retrieves it, and uses that data to render that portion of the Handlebars template. If I had to retrieve this data before calling Handlebars' compile method, I would have to recursively iterate through a JSON document whose structure I don't know in advance, find all the instances of external schemas, retrieve them, and insert them into the JSON.

Basically, anytime a Handlebars template is used to render JSON schema data ([json-schema.org](http://json-schema.org/)), an async render method would be useful, because JSON schema always allows subparts of a schema to be defined externally.

**Comment by nknapp** [comment]
@dwhieb have you had a look at [bootprint-swagger](https://github.com/nknapp/bootprint-swagger) for the documentation-generator? It's almost what you describe (except that external schemas are not yet implemented, but that would be a great feature). If you have any feedback, please open an issue there.

**Comment by richardkazuomiller** [comment]
I have a use case where being able to use promises in helpers would be useful. I'm using handlebars to generate the HTML for my blog. In order to build valid structured data for each article, I need to get the dimensions I'm using for the article image. Right now, I'm doing it like this:

````
{{#imageSize post.frontMatter.previewImage}}
  <div itemprop="image" itemscope itemtype="https://schema.org/ImageObject">
    <meta itemprop="url" content="{{#staticResource ../post.frontMatter.previewImage}}{{/staticResource}}">
    <meta itemprop="width" content="{{width}}">
    <meta itemprop="height" content="{{height}}">
  </div>
{{/imageSize}}
````

The `imageSize` helper works because it reads the file synchronously, but ideally it should be able to do it asynchronously so rendering pages isn't slowed down by I/O. Also, doing this for an image at a URL, instead of on the filesystem, is impossible.

I'll look into using promised-handlebars and express-hbs but I think the ability to use promises in helper functions would be a great addition to Handlebars!

**Comment by n2liquid** [comment]
FWIW, I've been doing a lot of asynchronous HTML rendering using Hyperscript, `hyperscript-helpers`, ES7's `async/await`, and it's been a real joy. But of course, that solution only works for HTML. An async solution with Handlebars would let us generate other kinds of files asynchronously... However for HTML I think I'm never looking back!

**Comment by darrenoakey** [comment]
going to give express-hbs a try - but I think in 2018 its a strange thing not to support.  I know that there is a purist view that async stuff should not be done as part of the view but instead in some magical thing called a controller (as if MVC is somehow unarguably "correct") - but that's a little short sighted for two key reasons

a) external libraries - most people now are writing entirely with async/await - I think in my code more than 9 out of 10 functions are async... in some cases "just in case".   Not supporting an async function means all async libraries are suddenly completely inaccessible

b) generic controller functions.  I would argue that something like this:  

```html
    {{#query "select name, total from summary"}}
          <tr><td>{{this.name}}</td><td>{{this.total}}</td></tr>
    {{/query}}
```

is shorter, cleaner, easier to maintain, and basically superior in every conceivable way when compared with having a bespoke controller function that sticks that stuff into a variable and passes it into a template, which the template then has to know about and access.

**Comment by Atulin** [comment]
I see nothing's changed for the last 8 years? With everything and their mother getting full `async` support, even Node slowly adding top-level `await`, Hbs still offers no support for async helpers?

**Comment by nknapp** [comment]
@Atulin Honestly, I don't think this is going to happen soon, too. It may seem like a big project, but from 2016 to 2020, I was basically the only one who did active maintenance. Now, @jaylinski has taken over. I wanted to support im, but in the last months, I had other things to do in my free time... So he is more or less alone in this.

What I want to say: But there is no such thing as a "active development team" or any kind of  "planning". All we have done in the last years is answer to questions, fix bugs, especially security issues.

Unless this changes, there will be no native async support for Handlebars.

**Comment by Atulin** [comment]
Gotcha. I found a package that adds async helper support and it's working well, so I'll say "good enough"

**Comment by nknapp** [comment]
die you mean this? https://www.npmjs.com/package/promised-handlebars

**Comment by Atulin** [comment]
https://github.com/gastonrobledo/handlebars-async-helpers

**Comment by CheyenneForbes** [comment]
@Atulin if I was to submit a pull request based on some of the ideas in @gastonrobledo's project, is there a chance of you or @jaylinski accepting/merging it?

**Comment by Atulin** [comment]
I'm not a maintainer of Hbs, so it's not me you should be asking lol

**Comment by jaylinski** [comment]
@CheyenneForbes yes, there is a chance that we will accept an async-helpers-improvement (but only in version 5).

**Comment by toverux** [comment]
This should very much be supported. You won't need it 90% of the time sure, but you desperately need it when you do.
In our use case we allow clients to create templates in our platform. We provide an helper allowing to load an image from a library of images. We have to preload the whole library prior to calling handlebars, making it very inefficient (that's a big fat MongoDB query).
We had to finally use handlebars-async-helpers because we had a new use case : allow to manipulate images and rotate them with a helper. This is not doable synchronously and we couldn't rotate the whole library in advance because this would take minutes. Now we have an async helper that does it on-demand. Unfortunately, handlebars-async-helpers treats async tasks sequentially but at last we've been able to implement it. We're actually thinking to move away from handlebars if this is not implemented, because this raises serious issues for uses cases like ours.

**Comment by nknapp** [comment]
@Atulin About https://www.npmjs.com/package/promised-handlebars: 

I know, there hasn't been an update for over 6 years. And I am not actively working on it anymore. On the other hand, it has almost no dependencies and no known security issues as well. 

And it's a small library with a tight use-case so there is actually no need for an update imho.

Handlebars has not had a real update for years as well. When there is a major version bump I will make sure that promised-handlebars stays compatible.

If anyone wants to make a PR updating the documentation to remove `q`... Go for it.

I am not promising fast responses here. But if there is something critical, Handlebars maintainers can reach me through faster channels.

**Comment by nknapp** [comment]
In any case. Of course it would be more powerful to have it included in Handlebars directly. But I don't see the resources for implementing this.

**Comment by yosiasz** [comment]
if performance is the issue why not throttle it using required pagination and limit the data returned?
